### PR TITLE
Removed sources of panic from ports/glutin.

### LINKED ISSUE #10547: servo panics on exit on windows 10 x64
I'm using ea97c8905580e67e96ae00572d1c4e7af3d58736 windows 10 x64 (build 10586.104)

```
$ ./mach run -r "https://github.com"
ERROR:js::rust: Error at https://assets-cdn.github.com/assets/frameworks-17fc21583b56cf4c511436a15e7eb5f4f48d3d8c1006e654d3108ef67a707cae.js:6:23659: history is not defined

thread '<main>' panicked at 'called `Option::unwrap()` on a `None` value', ../src/libcore\option.rs:326
stack backtrace:
   0:          0x1f5255a - std::panicking::default_hook::_$u7b$$u7b$closure$u7d$$u7d$::hff309ab1d83ffd90
   1:          0x1f51a7e - std::panicking::default_hook::h08ad3bb09872855b
   2:          0x1f14383 - std::sys_common::unwind::begin_unwind_inner::h406d5f1a330b854b
   3:          0x1f15f1b - std::sys_common::unwind::begin_unwind_fmt::h57ea3fbee1a40196
   4:          0x1f44bab - rust_begin_unwind
   5:          0x1f66ca5 - core::panicking::panic_fmt::ha6b3c19493c123b3
   6:          0x1f677b0 - core::panicking::panic::h188c6c6a0fe5463c
   7:           0xfd5456 - glutin_app::window::Window::wait_events::h1f6001f3a4f44ef0
   8:           0x40341a - servo::main::h6924e5658370f6b5
   9:          0x1f5109d - std::sys_common::unwind::try::hb63b2f006f75d989
  10:          0x1f50eca - std::rt::lang_start::h12d///9b3310b48a726
  11:           0x4013b4 - _tmainCRTStartup
  12:           0x4014e7 - mainCRTStartup
  13:     0x7ffd86108101 - mz_zip_writer_add_to_central_dir.constprop.15
Servo exited with return value 101
```

### PR DESCRIPTION
**Pull request by asajeffrey**
Fixes #10547.

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/10910)

### CODE REVIEW

File: ports/glutin/window.rs [hunk 1]
```diff
@@ -320,7 +319,13 @@ impl Window {
 
     #[cfg(any(target_os = "macos", target_os = "windows"))]
     fn handle_next_event(&self) -> bool {
-        let event = self.window.wait_events().next().unwrap();
+        let event = match self.window.wait_events().next() {
+            None => {
+                warn!("Window event stream closed.");
+                return false;
+            },
```
Comment (samlh) [inline ports/glutin/window.rs]: Return true? #10547 implies that this can happen when the window is closing.

Comment (asajeffrey) [inline ports/glutin/window.rs]: I wasn't sure about what to do here, most likely the window is already closing, and I'm not sure what'll happen if we try to close it twice? Is it idempotent?

**Comment by KiChjang** [comment]
I can't make a call on what to do about the returned value here. r? @pcwalton

**Comment by asajeffrey** [comment]
Rebased.

**Comment by asajeffrey** [comment]
Rebased and bumped log to 0.3.6.

**Comment by asajeffrey** [comment]
Rebased.

I had a think about what the return value should be from `handle_next_event` if the event queue returns `None`. In this case, we must be in shutdown already, so sending a `Quit` event will most likely confuse the other threads, since this is the second quit event. So I think returning `false` is the thing to do here.

**Comment by samlh** [comment]
Not that I have any authority, but this looks good to me :).

**Comment by asajeffrey** [comment]
Rebased.

**Comment by metajack** [comment]
Hmm. I have a fix for this as well, but on the glutin side. See https://github.com/tomaka/glutin/pull/777. The difference is that the change to glutin will prevent wait_events() from returning None.

What happens that once glutin gets a WM_DESTROY event (to close the window) it shuts down its event loops which causes recv() to panic, but the iterator discards the error and just returns None.

This patch is fine though as it doesn't coflict; it may cease to be necessary in the near future.

**Comment by asajeffrey** [comment]
We probably want to remove sources of panic from ports/glutin anway, since it's one of our single points of failure, a panic in ports/glutin brings down the whole browser. I'm quite happy about having a belt-and-braces approach here.

**Comment by metajack** [comment]
The issue here is that by returning immediately, you now have a very hot loop. This whole loop was written with the assumption that it would block here, and now that it's gone as soon as WM_DESTROY comes in, we probably spike to 100% CPU until shutdown.

**Comment by asajeffrey** [comment]
@metajack ah, the problem is that there's nothing `handle_next_event` can do to indicate that event processing should finish at this point? It can return a boolean, but that just controls whether a quit event is sent. We could park the thread, but that's pretty nasty.

**Comment by metajack** [comment]
Event processing isn't finished, though. We're waiting for shutdown to fully propagate. In my patch wait_events() blocks until the WAKEUP message arrives, and then we'll pull the FinishedShuttingDown message out of the other event queue. Maybe we can block on that event queue if we're in shutdown?

The main issue here is we have two event queues and Rust has nothing like select().

**Comment by asajeffrey** [comment]
Moving the conversation about hot-spinning at shutdown to #11307.

**Pull request merged**